### PR TITLE
Implement IEquatable<T> in record types

### LINKED ISSUE #44920: Record types should implement IEquatable<T>
Record types should implement `IEquatable<T>`. The types already synthesize a `public virtual bool Equals(T other)` method.

### PR DESCRIPTION
**Pull request by cston**
Fixes #44920

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs [hunk 1]
```diff
@@ -319,6 +319,12 @@ internal override ImmutableArray<NamedTypeSymbol> GetDeclaredInterfaces(ConsList
                 }
             }
 
+            if (declaration.Kind == DeclarationKind.Record)
+            {
+                var type = DeclaringCompilation.GetWellKnownType(WellKnownType.System_IEquatable_T);
+                baseInterfaces.Add(type.Construct(this));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: Does this cause us to implement the interface "non-nullably?" Should we instead implement the interface with a nullable type argument, e.g. `record Rec : IEquatable<Rec?>`

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: I do not see mention of IEquatable in the [records spec](https://github.com/dotnet/csharplang/blob/master/proposals/records.md).

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: The parameter to `IEquatable<T>.Equals()` is `[AllowNull] T other`.

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: > I do not see mention of IEquatable in the records spec.

Good catch, thanks. I've updated https://github.com/dotnet/csharplang/pull/3552.

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: > The parameter to IEquatable<T>.Equals() is [AllowNull] T other.

So it is, thank you! It might be good to add tests that demonstrate the synthesized "strongly typed Equals" methods accept null arguments without warnings or runtime errors.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: >baseInterfaces.Add [](start = 16, length = 18)

What if the interface already specified explicitly?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs [hunk 2]
```diff
@@ -319,6 +319,12 @@ internal override ImmutableArray<NamedTypeSymbol> GetDeclaredInterfaces(ConsList
                 }
             }
 
+            if (declaration.Kind == DeclarationKind.Record)
+            {
+                var type = DeclaringCompilation.GetWellKnownType(WellKnownType.System_IEquatable_T);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:324]: >DeclaringCompilation.GetWellKnownType(WellKnownType.System_IEquatable_T); [](start = 27, length = 73)

Please properly handle missing and otherwise bad types.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs (hunk 1 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:325]: >baseInterfaces.Add [](start = 16, length = 18)

We should make sure we test error scenarios around this interface. Places like that usually try to locate matching syntax in the base list. Obviously, in this case there is no syntax to find. We should make sure compiler doesn't crash.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 3]
```diff
@@ -6290,6 +6315,181 @@ .maxstack  2
...
+@"using System;
+record A<T> : IEquatable<A<T>>
+{
+}
+record B : A<object>, IEquatable<A<object>>, IEquatable<B>;
+";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
+
+            AssertEx.Equal(new[] { "System.IEquatable<A<T>>" }, comp.GetMember<NamedTypeSymbol>("A").AllInterfacesNoUseSiteDiagnostics.ToTestDisplayStrings());
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6388]: >AllInterfacesNoUseSiteDiagnostics [](start = 101, length = 33)

What do we return from Interfaces API?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 4]
```diff
@@ -6290,6 +6315,181 @@ .maxstack  2
...
+{
+    public virtual bool Equals(A<T> other) => false;
+}
+record B : A<object>, IEquatable<A<object>>, IEquatable<B>
+{
+    public override bool Equals(A<object> other) => false;
+    public virtual bool Equals(B other) => false;
+}";
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics();
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6407]: >comp.VerifyDiagnostics(); [](start = 12, length = 25)

Please run the code and observe expected methods called. Otherwise, how do we know that right methods are going to be hooked up?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 5]
```diff
@@ -6290,6 +6315,181 @@ .maxstack  2
...
+            var source =
+@"using System;
+record A<T> : IEquatable<A<T>>
+{
+    bool IEquatable<A<T>>.Equals(A<T> other) => false;
+}
+record B : A<object>, IEquatable<A<object>>, IEquatable<B>
+{
+    bool IEquatable<A<object>>.Equals(A<object> other) => false;
+    bool IEquatable<B>.Equals(B other) => false;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6425]: >bool IEquatable<B>.Equals(B other) => false; [](start = 4, length = 44)

Is this going to be a valid code without explicit ```IEquatable<B>``` in the base list?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 6]
```diff
@@ -6290,6 +6315,181 @@ .maxstack  2
...
+            var source =
+@"using System;
+record A<T> : IEquatable<A<T>>
+{
+    public virtual bool Equals(A<T> other) => false;
+}
+record B : A<object>, IEquatable<A<object>>, IEquatable<B>
+{
+    public override bool Equals(A<object> other) => false;
+    public virtual bool Equals(B other) => false;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6404]: >public virtual bool Equals(B other) => false; [](start = 4, length = 45)

Please add a test for a declaration like that, but without explicit ```IEquatable<B>``` in the base list.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6404]: >public virtual bool Equals(B other) => false; [](start = 4, length = 45)

Please also add a test for a scenario when this method is 
1) not public 
2) not virtual

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 7]
```diff
@@ -6290,6 +6315,181 @@ .maxstack  2
 }");
         }
 
+        [Fact]
+        public void IEquatableT_01()
+        {
+            var source =
+@"record A<T>;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6322]: >record A<T> [](start = 2, length = 11)

Please add a test for scenario when ```IEquatable<T>``` has unexpected API in it.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 1)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs (hunk 2 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol_Bases.cs:324]: Use-site errors are already reported for interfaces. See `RecordTests.IEquatableT_08`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs (hunk 5 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:6425]: Logged https://github.com/dotnet/roslyn/issues/45026.
